File: test/com/barrybecker4/optimization/optimizee/optimizees/problems/SubsetSumProblem.java
// Copyright by Barry G. Becker, 2013. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.optimization.optimizee.optimizees.problems;

import com.barrybecker4.optimization.Optimizer;
import com.barrybecker4.optimization.optimizee.optimizees.OptimizeeProblem;
import com.barrybecker4.optimization.parameter.ParameterArray;
import com.barrybecker4.optimization.parameter.VariableLengthIntArray;
import com.barrybecker4.optimization.strategy.OptimizationStrategyType;
import static com.barrybecker4.optimization.OptimizerTestCase.LOG_FILE_HOME;

/**
 * Determining if a set of numbers has a subset that sums to 0 is NP-Complete.
 * The only strategy that is guaranteed to find a solution if it exists is brute force search.
 * http://en.wikipedia.org/wiki/Subset_sum_problem
 *
 * @author Barry Becker
 */
public class SubsetSumProblem extends OptimizeeProblem {

    private SubsetSumVariation variation_;


    /** constructor */
    public SubsetSumProblem(SubsetSumVariation variation) {
        variation_ = variation;
    }

    @Override
    public String getName() {
        return "Subset Sum: " + variation_.name();
    }

    /**
     * we evaluate directly not by comparing with a different trial.
     */
    @Override
    public boolean evaluateByComparison() {
        return false;
    }

    /**
     * Use the cost matrix for the TSP variation to determine this.
     * @return fitness value
     */
    @Override
    public double evaluateFitness(ParameterArray a) {
        return variation_.evaluateFitness(a);
    }

    @Override
    public ParameterArray getExactSolution() {
        return variation_.getExactSolution();
    }

    @Override
    public ParameterArray getInitialGuess() {
        return variation_.getInitialGuess();
    }

    @Override
    public double getFitnessRange() {
        return variation_.getFitnessRange();
    }

    /**
     * This finds the solution for the above optimization problem.
     */
    public static void main(String[] args) {
        OptimizeeProblem problem = new SubsetSumProblem(SubsetSumVariation.NO_SOLUTION);

        Optimizer optimizer =
                new Optimizer(problem, LOG_FILE_HOME + "domSet_optimization.txt");

        ParameterArray initialGuess = problem.getInitialGuess();
        System.out.println("initial guess="+ initialGuess + " all=" + ((VariableLengthIntArray)initialGuess).getMaxLength());

        ParameterArray solution =
                optimizer.doOptimization(OptimizationStrategyType.GLOBAL_HILL_CLIMBING,
                        initialGuess, problem.getFitnessRange());

        showSolution(problem, solution);
    }
}

File: scala-test/com/barrybecker4/optimization/optimizee/optimizees/problems/SubsetSumProblem.scala
// Copyright by Barry G. Becker, 2013-2018. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.optimization.optimizee.optimizees.problems

import com.barrybecker4.optimization.Optimizer
import com.barrybecker4.optimization.optimizee.OptimizerTestSuite.LOG_FILE_HOME
import com.barrybecker4.optimization.optimizee.optimizees.OptimizeeProblem
import com.barrybecker4.optimization.parameter.{NumericParameterArray, ParameterArray, ParameterArrayWithFitness, VariableLengthIntSet}
import com.barrybecker4.optimization.strategy.{GLOBAL_HILL_CLIMBING, OptimizationStrategyType}

import scala.util.Random


object SubsetSumProblem {

  /** This finds the solution for the above optimization problem. */
  def main(args: Array[String]): Unit = {
    val problem = new SubsetSumProblem(NO_SOLUTION)
    val optimizer = new Optimizer(problem, Some(LOG_FILE_HOME + "domSet_optimization.txt"))
    val initialGuess = problem.getInitialGuess
    //System.out.println("initial guess=" + initialGuess + " all=" + initialGuess.asInstanceOf[VariableLengthIntSet].getMaxLength)
    val solution = optimizer.doOptimization(GLOBAL_HILL_CLIMBING,
      initialGuess, problem.getFitnessRange, new Random(1))
    OptimizeeProblem.showSolution(problem, solution)
  }
}

/**
  * Determining if a set of numbers has a subset that sums to 0 is NP-Complete.
  * The only strategy that is guaranteed to find a solution if it exists is brute force search.
  * http://en.wikipedia.org/wiki/Subset_sum_problem
  * @author Barry Becker
  */
class SubsetSumProblem(var variation: SubsetSumVariation) extends OptimizeeProblem[NumericParameterArray] {

  override def getName: String = "Subset Sum: " + variation.getName

  /** Evaluate directly not by comparing with a different trial. */
  override def evaluateByComparison = false

  /** Use the cost matrix for the TSP variation to determine this.
    * @return fitness value
    */
  override def evaluateFitness(a: NumericParameterArray): Double = variation.evaluateFitness(a)

  override def getExactSolution: ParameterArrayWithFitness = variation.getExactSolution

  override def getInitialGuess: NumericParameterArray = variation.getInitialGuess

  override def getFitnessRange: Double = variation.getFitnessRange
}
